cookbook/rule/add_custom_action_rule_engine.rst

```diff
@@ -0,0 +1,429 @@
+         */
+        public function getOptions()
+        {
+            return $this->options;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function setOptions(array $options = [])
+        {
+            $this->options = $options;
+
+            return $this;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function getImpactedFields()
+        {
+            return [$this->getField()];
+        }
+
+        /**
+         * @return array
+         */
+        public function getAttributes()
+        {
+            return $this->attributes;
+        }
+
+        /**
+         * @return array
+         */
+        public function getPattern()
+        {
+            return $this->pattern;
+        }
+
+        /**
+         * @param array $attributes
+         *
+         * @return ProductAddActionInterface
+         */
+        public function setAttributes(array $attributes = [])
+        {
+            $this->attributes = $attributes;
+
+            return $this;
+        }
+
+        /**
+         * @param string $pattern
+         *
+         * @return ProductAddActionInterface
+         */
+        public function setPattern($pattern)
+        {
+            $this->pattern = $pattern;
+
+            return $this;
+        }
+    }
+
+We also need to create a denormalizer that will return our previous object that handle the data. It will convert the array into an object.
+
+.. code-block:: php
+
+    #/src/Acme/Bundle/CustomBundle/Denormalizer/ProductRule/PatternActionDenormalizer.php
+    <?php
+
+    namespace Acme\Bundle\CustomBundle\Denormalizer\ProductRule;
+
+    use Acme\Bundle\CustomBundle\Model\ProductPatternActionInterface;
+    use Symfony\Component\Serializer\Normalizer\DenormalizerInterface;
+
+    class PatternActionDenormalizer implements DenormalizerInterface
+    {
+        /** @var string */
+        protected $patternActionClass;
+
+        /**
+         * @param string $patternActionClass should implement
+         *     \PimEnterprise\Component\CatalogRule\Model\ProductPatternActionInterface
+         */
+        public function __construct($patternActionClass)
+        {
+            $this->patternActionClass = $patternActionClass;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function denormalize($data, $class, $format = null, array $context = [])
+        {
+            return new $this->patternActionClass($data);
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function supportsDenormalization($data, $type, $format = null)
+        {
+            return isset($data['type']) && ProductPatternActionInterface::ACTION_TYPE === $data['type'];
+        }
+    }
+
+For our example we need to create an `ExistingAttributeValidator` that will check if the attributes provided in the rule file exists or not. It will raise a violation and skip the item in the case the attribute does not exist.
```

Comment:
"attributes provided in the rule file exist" or "attribute provided in the rule file exists"